### ISSUE TITLE
NPC glitches on paper

### ISSUE THREAD

**Issue by SlimeDog** [CitizensDev/Citizens2#2635]
```
> version
[15:24:40 INFO]: This server is running Paper version git-Paper-186 (MC: 1.17.1) (Implementing API version 1.17.1-R0.1-SNAPSHOT) (Git: 72ce684)
You are running the latest version
Previous version: git-Paper-184 (MC: 1.17.1)
>version Citizens
[15:15:41] [Server thread/INFO]: Citizens version 2.0.28-SNAPSHOT (build 2360)
[15:15:41] [Server thread/INFO]: Website: http://www.citizensnpcs.co
[15:15:41] [Server thread/INFO]: Author: fullwall
>version Sentinel
[15:26:29] [Server thread/INFO]: Sentinel version 2.3.5-SNAPSHOT (build 424)
[15:26:29] [Server thread/INFO]: Author: mcmonkey
```

The paper fix related to `Implement-Chunk-Priority-Urgency-System-for-Chunks" (Fixes #5980)` found in paper-186 (and paper-185, and probably paper-183 (not verified)) does not play well with Citizens. The issue does not exist in paper-184, or on comparable versions of Spigot.

On server start-up:
```
[15:22:44] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/801, uuid='62fe6737-fffe-4170-b58f-43aeac0dd8d2', l='ServerLevel[world_city]', x=-558.50, y=65.00, z=82.50, cpos=[-35, 5], tl=0, v=false, rR=null]
[15:22:44] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/802, uuid='e9b4024e-dea6-479b-bc1b-14ea1a8f6881', l='ServerLevel[world_city]', x=-559.50, y=65.00, z=75.50, cpos=[-35, 4], tl=0, v=false, rR=null]
```
All logged references to Citizens, Sentinel, and the WARNs:
```
[15:20:40] [Server thread/WARN]: Legacy plugin PluginManager v2.6 does not specify an api-version.
[15:20:40] [Server thread/WARN]: Legacy plugin WhatPlugin v1.0-SNAPSHOT (build #3) does not specify an api-version.
[15:20:40] [Server thread/WARN]: Legacy plugin EntityList v1.0 does not specify an api-version.
[15:20:40] [Server thread/WARN]: Legacy plugin RedstoneDetector v0.3 does not specify an api-version.
[15:20:40] [Server thread/WARN]: Legacy plugin IronDoors v1.6 does not specify an api-version.
[15:20:40] [Server thread/WARN]: Legacy plugin TakeAim v0.4.1 does not specify an api-version.
[15:20:40] [Server thread/WARN]: Legacy plugin DelayJoin v1.0-SNAPSHOT does not specify an api-version.
[15:20:40] [Server thread/WARN]: Legacy plugin UsernameHistory v1.2 does not specify an api-version.
[15:20:40] [Server thread/WARN]: Legacy plugin WDLCompanion v1.2.0 does not specify an api-version.
[15:20:40] [Server thread/WARN]: Legacy plugin AggressiveAnimals v2.6 does not specify an api-version.
[15:20:40] [Server thread/WARN]: Legacy plugin Sounds v1.3.2a does not specify an api-version.
[15:20:41] [Server thread/WARN]: Legacy plugin AcceptTheRules v7.5 does not specify an api-version.
[15:20:41] [Server thread/WARN]: Legacy plugin HungerKeeperPlus v1.4.1 does not specify an api-version.
[15:20:41] [Server thread/WARN]: Legacy plugin QuickBoard v4.4.1 does not specify an api-version.
[15:20:41] [Server thread/WARN]: Legacy plugin KeyCommandBind v3.6 does not specify an api-version.
[15:20:42] [Server thread/WARN]: Legacy plugin BiomeFinder v1.3.2 does not specify an api-version.
[15:20:42] [Server thread/WARN]: Legacy plugin ServerTutorialPlus v1.22.1 does not specify an api-version.
[15:20:42] [Server thread/WARN]: Legacy plugin AdvancedReplay v1.6.6 does not specify an api-version.
[15:20:42] [Server thread/WARN]: Legacy plugin MessageHacker v1.3.0 does not specify an api-version.
[15:20:45] [Server thread/INFO]: [Citizens] Loading Citizens v2.0.28-SNAPSHOT (build 2360)
[15:20:45] [Server thread/INFO]: [Sentinel] Loading Sentinel v2.3.5-SNAPSHOT (build 424)
[15:20:46] [Server thread/INFO]: [CombatLogX] Loading expansion 'Citizens Compatibility v16.4'...
[15:20:58] [Server thread/WARN]: **** SERVER IS RUNNING IN OFFLINE/INSECURE MODE!
[15:21:19] [Server thread/INFO]: [Citizens] Enabling Citizens v2.0.28-SNAPSHOT (build 2360)
[15:21:19] [Server thread/INFO]: [Sentinel] Enabling Sentinel v2.3.5-SNAPSHOT (build 424)
[15:21:19] [Server thread/INFO]: [Sentinel] Sentinel loading...
[15:21:19] [Server thread/INFO]: [Sentinel] Running on java version: 16.0.2
[15:21:20] [Server thread/INFO]: [Sentinel] Vault linked! Group targets will work.
[15:21:20] [Server thread/INFO]: [Sentinel] Sentinel found WorldGuard! Adding support for it!
[15:21:20] [Server thread/INFO]: [Sentinel] Sentinel loaded!
[15:21:20] [Server thread/INFO]: [Disease] Citizens plugin Found
[15:21:20] [Server thread/WARN]: [Denizen] Debug is disabled in the Denizen config. This is almost always a mistake, and should not be done in the majority of cases.
[15:21:21] [Server thread/WARN]: [Multiverse-Core] "Multiverse-Core v4.3.1-SNAPSHOT-b857" has registered a listener for org.bukkit.event.entity.EntityCreatePortalEvent on method "public void com.onarandombox.MultiverseCore.listeners.MVPortalListener.entityPortalCreate(org.bukkit.event.entity.EntityCreatePortalEvent)", but the event is Deprecated. "Server performance will be affected"; please notify the authors [dumptruckman, Rigby, fernferret, lithium3141, main--].
[15:22:00] [Server thread/INFO]: [Shopkeepers] Citizens found, enabling NPC shopkeepers.
[15:22:02] [Server thread/INFO]: [PhatLoots] Listening for Citizens NPC deaths
[15:22:03] [Server thread/INFO]: [SpiderNest][WARNING]World world_whitedog_the_end not found. Make sure it exists and is correctly spelled!
[15:22:03] [Server thread/INFO]: [SpiderNest][WARNING]World world_yellowdog_the_end not found. Make sure it exists and is correctly spelled!
[15:22:04] [Server thread/INFO]: [Quests] Successfully linked Quests with Citizens 2.0.28-SNAPSHOT (build 2360)
[15:22:07] [Server thread/INFO]: [CombatLogX] Enabling expansion 'Citizens Compatibility v16.4'...
[15:22:07] [Server thread/INFO]: [CombatLogX] [Citizens Compatibility] Successfully found a dependency: Citizens v2.0.28-SNAPSHOT (build 2360)
[15:22:07] [Server thread/INFO]: [CombatLogX] [Citizens Compatibility] Successfully found a dependency: Sentinel v2.3.5-SNAPSHOT (build 424)
[15:22:26] [Server thread/INFO]: [AACAdditionPro] Sentinel.BetterSprinting has been disabled in the config.
[15:22:26] [Server thread/INFO]: [AACAdditionPro] Sentinel.5Zig has been disabled in the config.
[15:22:26] [Server thread/INFO]: [AACAdditionPro] Sentinel.LabyMod has been disabled in the config.
[15:22:26] [Server thread/INFO]: [AACAdditionPro] Sentinel.Schematica has been disabled in the config.
[15:22:26] [Server thread/INFO]: [AACAdditionPro] Sentinel.Vape is not compatible with your server version.
[15:22:26] [Server thread/INFO]: [AACAdditionPro] Sentinel.WorldDownloader has been disabled in the config.
[15:22:26] [Server thread/INFO]: [AACAdditionPro] Sentinel.Custom.Forge has been disabled in the config.
[15:22:26] [Server thread/INFO]: [AACAdditionPro] Sentinel.Custom.LiteLoader has been disabled in the config.
[15:22:26] [Server thread/INFO]: [AACAdditionPro] Sentinel.Custom.PixelClient has been disabled in the config.
[15:22:26] [Server thread/INFO]: [AACAdditionPro] Sentinel.Custom.PXMod has been disabled in the config.
[15:22:26] [Server thread/INFO]: [AACAdditionPro] Sentinel.Custom.Rift has been disabled in the config.
[15:22:26] [Server thread/INFO]: [AACAdditionPro] Sentinel.Custom.Winterware has been disabled in the config.
[15:22:27] [Server thread/INFO]: [MobHunting] Enabling compatibility with Citizens2 (2.0.28-SNAPSHOT (build 2360))
[15:22:29] [Craft Scheduler Thread - 5/INFO]: [ItemJoin] Hooked into { Multiverse-Core, Multiverse-Inventories, WorldGuard, PlaceholderAPI, ProtocolLib, Citizens, Vault }
[15:22:31] [Server thread/WARN]: [ViaVersion] ViaVersion does not have any compatible versions for this server version!
[15:22:31] [Server thread/WARN]: [ViaVersion] Please remember that ViaVersion only adds support for versions newer than the server version.
[15:22:31] [Server thread/WARN]: [ViaVersion] If you need support for older versions you may need to use one or more ViaVersion addons too.
[15:22:38] [Server thread/INFO]: [Citizens] Loaded 32 NPCs.
[15:22:39] [Server thread/WARN]: Can't keep up! Is the server overloaded? Running 9532ms or 190 ticks behind
[15:22:44] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/801, uuid='62fe6737-fffe-4170-b58f-43aeac0dd8d2', l='ServerLevel[world_city]', x=-558.50, y=65.00, z=82.50, cpos=[-35, 5], tl=0, v=false, rR=null]
[15:22:44] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/802, uuid='e9b4024e-dea6-479b-bc1b-14ea1a8f6881', l='ServerLevel[world_city]', x=-559.50, y=65.00, z=75.50, cpos=[-35, 4], tl=0, v=false, rR=null]
```
The two NPCs exist in Citizens/saves.yml, but nowhere else on the server (except the WARN log entries:
```
./plugins/Citizens/saves.yml:    uuid: e9b4024e-dea6-479b-bc1b-14ea1a8f6881
./plugins/Citizens/saves.yml:    uuid: 62fe6737-fffe-4170-b58f-43aeac0dd8d2
```
And they exist in-game. But they do not have complete information:
```
> npc select 15
[15:33:20 INFO]: Selected Guinness (ID 15).
> npc
[15:33:23 INFO]: =====[ Guinness ]=====
[15:33:23 INFO]:     ID: 15
[15:33:23 INFO]:     UUID: e9b4024e-dea6-479b-bc1b-14ea1a8f6881
[15:33:23 INFO]:     Type: WOLF
[15:33:23 INFO]:     Traits
[15:33:23 INFO]:      - sentinel
> npc select 16
[15:33:27 INFO]: Selected Murphy (ID 16).
> npc
[15:33:28 INFO]: =====[ Murphy ]=====
[15:33:28 INFO]:     ID: 16
[15:33:28 INFO]:     UUID: 62fe6737-fffe-4170-b58f-43aeac0dd8d2
[15:33:28 INFO]:     Type: WOLF
[15:33:28 INFO]:     Traits
[15:33:28 INFO]:      - sentinel
```
Normally, they would reveal their locations:
```
>npc select 15
[15:33:59] [Server thread/INFO]: Selected Guinness (ID 15).
>npc
[15:34:01] [Server thread/INFO]: =====[ Guinness ]=====
[15:34:01] [Server thread/INFO]:     ID: 15
[15:34:01] [Server thread/INFO]:     UUID: e9b4024e-dea6-479b-bc1b-14ea1a8f6881
[15:34:01] [Server thread/INFO]:     Type: WOLF
[15:34:01] [Server thread/INFO]:     Spawned at -560, 65, 74 in world world_city
[15:34:01] [Server thread/INFO]:     Traits
[15:34:01] [Server thread/INFO]:      - sentinel
>npc select 16
[15:34:06] [Server thread/INFO]: Selected Murphy (ID 16).
>npc
[15:34:07] [Server thread/INFO]: =====[ Murphy ]=====
[15:34:07] [Server thread/INFO]:     ID: 16
[15:34:07] [Server thread/INFO]:     UUID: 62fe6737-fffe-4170-b58f-43aeac0dd8d2
[15:34:07] [Server thread/INFO]:     Type: WOLF
[15:34:07] [Server thread/INFO]:     Spawned at -559, 65, 81 in world world_city
[15:34:07] [Server thread/INFO]:     Traits
[15:34:07] [Server thread/INFO]:      - sentinel
```

Presumably, Paper has broken something that Citizens or Sentinel needs. This may well be a paper bug, but I would need more information than I currently understand before opening an issue on paper.

[Paper changes](https://papermc.io/downloads):
186 [72ce684] Fix possible divide by 0 error with cursor updates (#6398)
185 [8d182e8] Revert "Revert "Remove "Implement-Chunk-Priority-Urgency-System-for-Chunks" (Fixes #5980)""
184 [622598f] Revert "Remove "Implement-Chunk-Priority-Urgency-System-for-Chunks" (Fixes #5980)"
183 [aa52bf9] Remove "Implement-Chunk-Priority-Urgency-System-for-Chunks" (Fixes #5980)

**Comment by mcmonkey4eva** [comment]
Per the Paper issue - looks to be not related to Citizens, and instead caused by a custom generator plugin performing unsafe async calls.

**Comment by SlimeDog** [comment]
Yep, thanks.

**Comment by SlimeDog** [comment]
That was the Paper hanging issue. Unless I am mistaken, the issues with Citizens remain.

**Comment by mcmonkey4eva** [comment]
`And they exist in-game. But they do not have complete information:`
That "incomplete information" is the NPCs not being spawned. Given the console message, It's possible somehow the entity gets duplicated, such that an entity copy is spawned and the real NPC spawn fails.

**Comment by SlimeDog** [comment]
OK. What more information can I provide?

**Comment by SlimeDog** [comment]
FWIW, `npc list -a` lists the expected (but informationally deficient) NPCs, and does not list any unexpected duplicate NPCs.

**Comment by SlimeDog** [comment]
This issue persists on Paper 1.17.1-196.

**Comment by SlimeDog** [comment]
The server has 32 enabled NPCs, mostly Sentinel wolves. On most server restarts, the two shown above arrive in the described condition; the remainder are fine. The two are the only wolf NPCs in world_city, although there are other NPCs in that world. Very occasionally, maybe once in 20 restarts, three more Sentinel wolves in another world (world_blackdog) arrive in the same condition. Extremely rarely, maybe once in a 100 restarts, all NPCs are correctly spawned.

This strongly suggests a timing issue. Is there a way to delay enabling Citizens/Denizen/Sentinel, in order to test that hypothesis?

**Comment by SlimeDog** [comment]
This issue persists on Paper 1.17.1-249 with
- Citizens .0.28-SNAPSHOT (build 2367)
- Sentinel version 2.3.5-SNAPSHOT (build 431)

As before, everything continues to work on Spigot 1.17.1-3236, with the same content (just a change of Minecraft variant).

**Comment by SlimeDog** [comment]
```
[21:21:40] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/590, uuid='c7113d5e-6f1b-41f0-9413-03cfc3bed436', l='ServerLevel[world_blackdog]', x=158.50, y=72.00, z=-297.50, cpos=[9, -19], tl=0, v=false, rR=null]
[21:21:40] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/591, uuid='f03a66b6-dc42-4c30-895e-52b6adb70956', l='ServerLevel[world_westeros]', x=2502.50, y=71.00, z=2856.60, cpos=[156, 178], tl=0, v=false, rR=null]
[21:21:40] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/592, uuid='9327b325-521b-4007-b2b6-3b4d325f22ea', l='ServerLevel[world_blackdog]', x=144.50, y=72.00, z=-302.50, cpos=[9, -19], tl=0, v=false, rR=null]
[21:21:40] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/593, uuid='f6f01b49-d9a0-4c97-8c22-7b365ab9373b', l='ServerLevel[world_blackdog]', x=144.50, y=72.00, z=-292.50, cpos=[9, -19], tl=0, v=false, rR=null]
[21:21:40] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/594, uuid='62fe6737-fffe-4170-b58f-43aeac0dd8d2', l='ServerLevel[world_city]', x=-558.50, y=65.00, z=82.50, cpos=[-35, 5], tl=0, v=false, rR=null]
```
With `debug-mode: true` the following may prove useful. These were the NPC that failed in this instance (it is not always the same ones).
```
[21:21:40] [Server thread/INFO]: [Citizens] Retrying spawn of 30 later, SpawnReason.PLUGIN. Was loaded true is loaded true
[21:21:40] [Server thread/INFO]: [Citizens] Stored 30 for respawn from NPCNeedsRespawnEvent
[21:21:40] [Server thread/INFO]: [Citizens] Retrying spawn of 28 later, SpawnReason.PLUGIN. Was loaded true is loaded true
[21:21:40] [Server thread/INFO]: [Citizens] Stored 28 for respawn from NPCNeedsRespawnEvent
[21:21:40] [Server thread/INFO]: [Citizens] Retrying spawn of 19 later, SpawnReason.PLUGIN. Was loaded true is loaded true
[21:21:40] [Server thread/INFO]: [Citizens] Stored 19 for respawn from NPCNeedsRespawnEvent
[21:21:40] [Server thread/INFO]: [Citizens] Retrying spawn of 18 later, SpawnReason.PLUGIN. Was loaded true is loaded true
[21:21:40] [Server thread/INFO]: [Citizens] Stored 18 for respawn from NPCNeedsRespawnEvent
[21:21:40] [Server thread/INFO]: [Citizens] Retrying spawn of 16 later, SpawnReason.PLUGIN. Was loaded true is loaded true
[21:21:40] [Server thread/INFO]: [Citizens] Stored 16 for respawn from NPCNeedsRespawnEvent
```
Is there additional information I can provide?

**Comment by SlimeDog** [comment]
This issue persists on Paper 1.17.1-266 with

- Citizens 2.0.28-SNAPSHOT (build 2371)
- Sentinel version 2.3.5-SNAPSHOT (build 431)

As before, everything continues to work on Spigot 1.17.1-3241, with the same content (just a change of Minecraft variant).

**Comment by fullwall** [comment]
I suspect it's because Citizens now uses the NPC uuid as the entity UUID, and is double spawning. This only happens with sentinel right?

**Comment by SlimeDog** [comment]
> I suspect it's because Citizens now uses the NPC uuid as the entity UUID, and is double spawning. This only happens with sentinel right?

Correct, in my (limited) environment, and only with some of my 21 sentinels in various worlds. As described above, it usually occurs with the same five, but sometimes with others, and (very occasionally) with none.

I have 11 other NPC, non-sentinels, which have never exhibited this problem.

**Comment by SlimeDog** [comment]
I'm mystified why this occurs only on Paper. Isn't UUID-based NPC identification is used on Spigot as well?

**Comment by SlimeDog** [comment]
This issue persists on Paper 1.17.1-275 with
- Citizens 2.0.28-SNAPSHOT (build 2382)
- Sentinel version 2.3.5-SNAPSHOT (build 432)

Consequently, Paper cannot be used in production, since various sentinels are glitched. As before, everything continues to work on Spigot 1.17.1-3242, with the same content (just a change of Minecraft variant).

**Comment by fullwall** [comment]
I appreciate the posts but I'm not sure if this is a Sentinel issue or a Citizens issue... at least I haven't been able to replicate it so far except when I am testing things.

**Comment by SlimeDog** [comment]
Should I cross-post to Sentinel?

**Comment by mcmonkey4eva** [comment]
`Is there additional information I can provide?`
Things that could be helpful to post:
- Full unaltered `logs/latest.log`. Make sure to resolve any standing errors in the log wherever possible, notably including `[Denizen] Debug is disabled in the Denizen config. This is almost always a mistake, and should not be done in the majority of cases.` visible in your earlier post
- Full unaltered `Citizens/saves.yml`

If possible and willing to spend the time on it, try to whittle your server down to the minimum set of content needed to replicate the issue (on a test server not your live server of course). Mostly meaning, removing all unrelated plugins, and if possible also resetting your server and plugin configs and saves, and remaking NPCs.
See also https://wiki.citizensnpcs.co/Plugin_Conflict regarding strategy to do that efficiently. tl;dr: do a binary search, remove half of what's on the server at a time and see if the problem persists or not - if not, readd that content and try removing the other half - repeat until you have a server where if anything gets removed or reset, the problem goes away. Then, of course, post a list of all the things that have to be exactly as they are for the problem to persist.

**Comment by SlimeDog** [comment]
Please find gzipped tarball at https://whitedog.straight8.com/drops/test-1.gz `(tar -xzf test-1.gz)` which is a minimal testbed to produce the errors. 
- size: 541M
- md5sum: ea05d42d310d3630004945b5dae90c02

Please advise when you have downloaded it, so that I can delete it from the web server.

The test environment is
```
$ cat /etc/centos-release
CentOS Linux release 7.9.2009 (Core)
$ java -version
java version "17" 2021-09-14 LTS
Java(TM) SE Runtime Environment (build 17+35-LTS-2724)
Java HotSpot(TM) 64-Bit Server VM (build 17+35-LTS-2724, mixed mode, sharing)
```
```
> version
[09:24:38 INFO]: This server is running Paper version git-Paper-278 (MC: 1.17.1) (Implementing API version 1.17.1-R0.1-SNAPSHOT) (Git: d0a2193)
You are running the latest version
```


The plugin load has been reduced to three
```
plugins//Citizens-2.0.28-b2382.jar
plugins//Multiverse-Core-4.3.2-b862.jar
plugins//Sentinel-2.3.5-b432.jar
```
Multiverse is required, since the NPCs of interest are in different worlds. I pared the UNIVERSE/ subdirectory to only the necessary worlds (our standard configuration includes many more)
```
UNIVERSE/world
UNIVERSE/world_blackdog
.UNIVERSE/world_city
UNIVERSE/world_nether
UNIVERSE/world_the_end
```
There are two commands in the bin/ subdirectory
- bin/mc-start
- bin/mc-errors

I left our standard `bin/mc-start` intact, but deleted its progeny, so some diagnostic output will be generated ("cannot execute ...", see below), which may be safely ignored. The script starts a Paper server in a `screen` session. If you prefer to run the server directly, execute `bin/mc-start --test-invocation` for the necessary java command line. The script may also be invoked with `bin/mc-start --verbose` if you want details about what it does.

I left `plugins/Citizens/saves.yml` intact, so some errors will be generated due to missing traits for other NPCs, which may be safely ignored.

The expected output is:
```
$ bin/mc-start 
         > Cannot execute bin/mc-start-pre
         > Cannot execute bin/mc-start-pre-temp
Use mcs to join the screen session, ^Ad to leave it
.........|.........|.........| [30 seconds]
         > Cannot execute bin/mc-start-post
         > Cannot execute bin/mc-start-post-temp
[08:30:30] [Server thread/ERROR]: [Citizens] The trait triggers failed to load for NPC ID: 5.
[08:30:30] [Server thread/ERROR]: [Citizens] The trait assignment failed to load for NPC ID: 5.
[08:30:30] [Server thread/ERROR]: [Citizens] The trait assignment failed to load for NPC ID: 11.
[08:30:30] [Server thread/ERROR]: [Citizens] The trait triggers failed to load for NPC ID: 11.
[08:30:30] [Server thread/ERROR]: [Citizens] The trait triggers failed to load for NPC ID: 17.
[08:30:30] [Server thread/ERROR]: [Citizens] The trait assignment failed to load for NPC ID: 17.
$ mc-errors 
[08:30:30] [Server thread/ERROR]: [Citizens] The trait triggers failed to load for NPC ID: 5.
[08:30:30] [Server thread/ERROR]: [Citizens] The trait assignment failed to load for NPC ID: 5.
[08:30:30] [Server thread/ERROR]: [Citizens] The trait assignment failed to load for NPC ID: 11.
[08:30:30] [Server thread/ERROR]: [Citizens] The trait triggers failed to load for NPC ID: 11.
[08:30:30] [Server thread/ERROR]: [Citizens] The trait triggers failed to load for NPC ID: 17.
[08:30:30] [Server thread/ERROR]: [Citizens] The trait assignment failed to load for NPC ID: 17.
[08:30:31] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/251, uuid='c7113d5e-6f1b-41f0-9413-03cfc3bed436', l='ServerLevel[world_blackdog]', x=158.50, y=72.00, z=-297.50, cpos=[9, -19], tl=0, v=false, rR=null]
[08:30:31] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/252, uuid='9327b325-521b-4007-b2b6-3b4d325f22ea', l='ServerLevel[world_blackdog]', x=144.50, y=72.00, z=-302.50, cpos=[9, -19], tl=0, v=false, rR=null]
[08:30:31] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/253, uuid='f6f01b49-d9a0-4c97-8c22-7b365ab9373b', l='ServerLevel[world_blackdog]', x=144.50, y=72.00, z=-292.50, cpos=[9, -19], tl=0, v=false, rR=null]
[08:30:31] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/254, uuid='62fe6737-fffe-4170-b58f-43aeac0dd8d2', l='ServerLevel[world_city]', x=-558.50, y=65.00, z=82.50, cpos=[-35, 5], tl=0, v=false, rR=null]
[08:30:31] [Server thread/WARN]: UUID of added entity already exists: EntityWolfNPC['Wolf'/255, uuid='e9b4024e-dea6-479b-bc1b-14ea1a8f6881', l='ServerLevel[world_city]', x=-559.50, y=65.00, z=75.50, cpos=[-35, 4], tl=0, v=false, rR=null]
```
As you can see, five sentinel NPCs report errors.

**Comment by mcmonkey4eva** [comment]
## The Test File

Downloaded and confirmed the errors do show.
For obvious security reasons, I copied only your world files and configs, and redownloaded Paper and plugins from source. Also had to change some configs and file locations (not sure what your "UNIVERSES" folder is, but that's not how worlds work in a normal server, and none of the config files seemed to be related to why you have that folder?)

Also, you might want to look into using LetsEncrypt for your webserver, the https link you posted doesn't work as it's self-signed - luckily I could just remove the `s` to download direct over http.

Testing with this file, it seems like it only happens when Multiverse is involved. Simply removing multiverse and changing the default world to `world_city` is sufficient to remove the errors.
Keeping multiverse installed and switching the default world to `world_city` causes NPCs in `world` to show the error (ie a world that is now loaded by Multiverse).
This seems consistent with your own log references - the error only appears in Multiverse worlds.

## Analysis

The only reason Sentinel seems to be involved is because Sentinel automatically tries to respawn entities during startup if needed - something went wrong in Citizens own native spawning, and then later Sentinel tries to spawn the NPC again, and that's where the message comes from - it is genuinely trying to spawn the same entity twice. However, of course, the Citizens spawn method of course doesn't allow spawning an already-spawned NPC, so the fact that it tries to spawn again anyway indicates clearly that the first spawn went horribly wrong in some way.

Disabling respawn on an affected wolf NPC makes it seem to work fine.

## Solution/Workaround

While this appears to be an underlying issue within Paper and/or Multiverse, I believe a valid workaround would be altering this line:
https://github.com/CitizensDev/Citizens2/blob/3d3de6576df4bc567966c9b3f5a77560bd3175e2/main/src/main/java/net/citizensnpcs/npc/CitizensNPC.java#L240
to instead only check `getEntity() != null`

Tested making that change locally, and now the NPC spawns perfectly every time, and if debug is enabled merely shows the expected message `[Citizens] Tried to spawn 4 while already spawned. SpawnReason.PLUGIN`

## Further Analysis Of Workaround

The change this makes is of course ignoring the `NMS.isValid(` check made within `isSpawned`.
Internally, that check consists of:
```java
        Entity handle = getHandle(entity);
        return handle.valid && handle.isAlive();
```
For the sake of testing, I added back the isSpawned call and inside isValid added the line:
`Messaging.debug("NPC handle ", handle, " valid? ", handle.valid, " alive? ", handle.isAlive());`
and got the following output before a duplicate entity warning:
`[17:56:56] [Server thread/INFO]: [Citizens] NPC handle  EntityWolfNPC['Butch'/6, uuid='655b05fd-05ca-45ba-bc67-9b710b57d26d', l='ServerLevel[world]', x=-14.50, y=64.00, z=14.50, cpos=[-1, 0], tl=0, v=false, rR=null]  valid?  false  alive?  true`
As can be seen: `handle.valid` is `false`, despite `isAlive()` being `true` and of course the entity itself existing.
That's the limit of my current ability to track/debug the issue. Something in Multiverse and Paper is conflicting in such a way to allow the entity to be spawned in such a way that `valid = false` despite it being a fully valid and functional entity.
All other details are simply side effects of this initial error.
(That being false caused the duplicate spawn to occur, the duplicate spawn caused the entity to lose its link to the first spawn and thus seemingly be unspawned as the second spawn didn't go through, etc)

**Comment by SlimeDog** [comment]
> not sure what your "UNIVERSES" folder is, but that's not how worlds work in a normal server, and none of the config files seemed to be related to why you have that folder?

./UNIVERSE/ is the worlds subfolder. This is standard practice to move to world subdirectories from the main directory.
`java ... --universe ./UNIVERSE`

> you might want to look into using LetsEncrypt for your webserver

I do use LetsEncrypt for my real web server. Apologies for giving you an `https` for my auxiliary server. Glad you could figure it out.

Thanks for the detailed analysis. I look forward to a published fix/workaround. Feel free to share the analysis with [Paper](https://github.com/PaperMC/Paper/issues/6405) and Multiverse to resolve the issue at the root cause.

**Comment by SlimeDog** [comment]
Since all my sentinels are invulnerable, I can safely set `respawnTime: -1` as a workaround.

**Comment by mcmonkey4eva** [comment]
I'd recommend setting `0` not `-1` just to be safe (0 means no respawn, -1 means auto-delete on death)

**Comment by SlimeDog** [comment]
Thanks. Done.

**Comment by mcmonkey4eva** [comment]
The workaround I recommended was added via https://github.com/CitizensDev/Citizens2/commit/18423134fafd858d031a84f5bda97c923b9738b6 for Citizens build 2383+ from https://ci.citizensnpcs.co/job/Citizens2/

Assuming the error cannot be replicated with this workaround present, this issue can be closed on the Citizens end of things at least, though ideally Paper and/or Multiverse should still fix the underlying issue.

**Comment by SlimeDog** [comment]
Confirmed fixed with Paper 1.17.1-279 and Citizens 2.0.28-SNAPSHOT (build 2383)
Thanks.
